Adding web component Tabs, Tab and TabPanel

bearcat-msft opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] Code is up-to-date with the `master` branch
* [ ] Your changes are covered by tests (if possible)
* [ ] You've run `yarn change` locally


PR flow tips:
* [ ] Try to start with a Draft PR
* [ ] Once you're ready (ideally the pipeline is passing) promote your PR to Ready for Review. This step will auto-assign reviewers for your PR.
-->

## Previous Behavior

FASTTabs and FASTTab render the activeIndicator as siblings to the tabs. The tab indicator rendering, dom structure, and animations do not match the behavior of the Fluent React Tabs.

## New Behavior

Extends FASTTab and FASTTabs. Adds new behavior for active indicator. Active indicator animations, DOM rendering, and logic follows the patterns set in the Fluent React TabList control.


## Notes about tab template and tab resize on hover and bold text

One problem encountered during the development of this PR was the rendering of tab content bold text on selection. When rendering a list of variable sized horizontal elements any size change of the text content will cause the list of elements to move slightly. Css properties like flex box, fit-content, or css grid will try to recalculate the element width when the inner content size changes (unless fixed element size is specified). The only way around this is to somehow set the width of the content container to a fixed size. Here are a few discoveries about rendering lists of horizontal elements:

- if the element has a fixed width, changing the width of the content will not cause the size of the container to change (so long as the content change does not exceed the size of the container)
- if an element has a fixed width and long content is inserted, the height of the content will compensate for the content size
- flex box, css grid, and fit content all resize the box width when the content size changes (like in the case of applying bold or semibold text)
- _KEY TAKEAWAY_: if variable sized horizontal elements are required, either the design must be okay with layout shift, or a way of adding the extra size to account for the change in size on hover must be provided (see below for how this was approached). No matter which css method is used, if it calculates the width based on content and the width is changing a layout shift will occur.

here is a breakdown of different methods for breaking horizontal layouts that I tried - https://stackblitz.com/edit/js-qxteyt?file=index.js,index.html,style.css&file=index.html,index.js

The fluent guidelines for the TabList allow for both variable sized tabs and bold text (ie size change) on hover. This PR adds a hidden pseudo element that renders invisible bold text. This element is always present in the UI, so when the user hovers the bold text that is applied does not add extra space and does not cause a layout shift.

fabricteam commented:
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/71c3165c70d04f3b7e65a9c30eed058a1c7eec6d'>71c3165c70d04f3b7e65a9c30eed058a1c7eec6d</a></sub>
<div id="bundleSizeComment9423"/>

halerankin reviewed (commented):
This looks great!

chrisdholt commented on packages/web-components/src/tabs/tab-list/tab-list.ts:
```diff
@@ -0,0 +1,82 @@
+ * @class TabList component
+ * @public
+ */
+export class TabList extends FASTTabs {
```
So, I don't know that extending here with the name `TabList` makes sense given that this is the _whole_ thing. The tablist does not include tabpanels - this class from FAST maps explicitly to the APG `Tabs` Component. A tablist is constructed, but I don't think it's going to be clear or correct to convolute these two simply because the React implementation doesn't implement panels (from what I am grokking).

If we'd like to match the naming, you probably need to change the implementation.

bearcat-msft replied:
So what do you think? Rename to Tabs and talk about the differences here in the readme?

chrisdholt replied:
If we want to keep the FAST Foundation code, I would rename to Tabs and document. Otherwise, we can implement according to the React implementation.

chrisdholt commented on packages/web-components/src/tabs/tab-list/tab-list.ts:
```diff
@@ -0,0 +1,82 @@
+ * @public
+ */
+export class TabList extends FASTTabs {
+  @attr appearance?: 'subtle' | 'transparent';
```
Let's please create const objects and types for these :)

chrisdholt commented on packages/web-components/src/tabs/tab-list/tab-list.ts:
```diff
@@ -0,0 +1,82 @@
+export class TabList extends FASTTabs {
+  @attr appearance?: 'subtle' | 'transparent';
+
+  @attr({ converter: booleanConverter })
+  disabled?: boolean;
```
You can use `mode: "boolean"` instead of importing and including that converter.

chrisdholt commented on packages/web-components/src/tabs/tab-list/tab-list.ts:
```diff
@@ -0,0 +1,82 @@
+  @attr({ converter: booleanConverter })
+  disabled?: boolean;
+
+  @attr size?: 'small' | 'medium' | 'large' = 'medium';
```
For all typings in here, please enumerate that they are public, protected, or private. All attrs will be public

chrisdholt commented on packages/web-components/src/tabs/tab-list/tab-list.ts:
```diff
@@ -0,0 +1,82 @@
+    }
+  }
+
+  // Remove this method after updates to Fast Tabs Control are completed
```
Do we have a task tracking whatever this is in reference to? (What is it in reference to?)

bearcat-msft replied:
removed the code this was in reference to

chrisdholt commented on packages/web-components/src/tabs/tab-list/tab-list.ts:
```diff
@@ -0,0 +1,82 @@
+
+      /* 
+      after setting all tabs to disabled the FAST tabs control 
+      sets showActiveIndicator to false. Fluent guidelines have the tab indicator always showing if the tab is active.
```
I think we discussed this offline, but our current hope/approach is to remove the prpescriptive `activeIndiciator` and all of the JS because it hoists presentational details to the whole component. Were we not able to do this without using that element? I believe you mentioned that the actual current v9 implementation uses a pseudo + animation.

chrisdholt commented on packages/web-components/src/tabs/index.ts:
@bearcat-msft all our other files are a flat structure - if you don't mind let's break these each out into individual files and then exporting these from the root index file.

chrisdholt commented on packages/web-components/src/tabs/tabs/tabs.options.ts:
```diff
@@ -0,0 +1,31 @@
+
+export type TabListOrientation = ValuesOf<typeof TabListOrientation>;
+
+export interface TabData {
+  id: string;
+  x: number;
+  y: number;
+  height: number;
+  width: number;
+}
```
From what I saw this was internals, do we need to export an interface to consumers for this?

chrisdholt reviewed (changes requested):
I'd like us to look a bit more on how extensive the code is for the animation _within_ the `Tab` portion of the component. We're looking up the tree there, but in Tabs we then pass information...It seems like we could significantly improve this and ensure that we manage it all from one place. If we know the active tab in `Tabs`, I don't think we need to know in the Tab whether it is active. Similarly, we can style the active tab from *Tabs*. I'd like us to look at having the animation code in a singular place, `Tabs`, because I think we can drastically simplify this.

bearcat-msft commented on packages/web-components/src/tabs/tab/tab.styles.ts:
```diff
@@ -0,0 +1,126 @@
+    color: ${colorNeutralForegroundDisabled};
+  }
+
+  :host([disabled]:hover) {
+    color: none;
+  }
+  :host([disabled]:hover)::before {
+    background-color: unset;
+  }
```
colorNeutralBackgroundDisabled does not map to the correct value.

bearcat-msft commented on packages/web-components/src/tabs/tab/tab.styles.ts:
```diff
@@ -0,0 +1,126 @@
+  :host(:focus) {
+    outline: none;
+  }
+  :host(.vertical) {
+    justify-content: start;
+    grid-column: 2;
+  }
+  :host(.vertical[aria-selected='true']) {
+    z-index: 2;
+  }
```
yes, these are throwbacks from foundation. I can move this css into the tabs parent and remove these class selectors.

bearcat-msft commented:
> I'd like us to look a bit more on how extensive the code is for the animation _within_ the `Tab` portion of the component. We're looking up the tree there, but in Tabs we then pass information...It seems like we could significantly improve this and ensure that we manage it all from one place. If we know the active tab in `Tabs`, I don't think we need to know in the Tab whether it is active. Similarly, we can style the active tab from _Tabs_. I'd like us to look at having the animation code in a singular place, `Tabs`, because I think we can drastically simplify this.

Thanks Chris, I think i addressed all the feedback for this. All the logic has been moved into tabs now.  One thing - for CSS, unless there is a technical reason to move the CSS from tab to tabs, doesn't it just add to the bloat of the CSS file for Tabs? There is a lot of css in there related to the animation and tabs attributes already. At what point are the concerns of Tab and Tabs separated?

chrisdholt commented:
> > I'd like us to look a bit more on how extensive the code is for the animation _within_ the `Tab` portion of the component. We're looking up the tree there, but in Tabs we then pass information...It seems like we could significantly improve this and ensure that we manage it all from one place. If we know the active tab in `Tabs`, I don't think we need to know in the Tab whether it is active. Similarly, we can style the active tab from _Tabs_. I'd like us to look at having the animation code in a singular place, `Tabs`, because I think we can drastically simplify this.
> 
> Thanks Chris, I think i addressed all the feedback for this. All the logic has been moved into tabs now. One thing - for CSS, unless there is a technical reason to move the CSS from tab to tabs, doesn't it just add to the bloat of the CSS file for Tabs? There is a lot of css in there related to the animation and tabs attributes already. At what point are the concerns of Tab and Tabs separated?

Thanks, I'll take a look as soon as I can - the cost of offloading CSS in shadow Dom is probably negligible compared to the amount of JS that was being added for each tab. If the trade-off is between CSS and JS, we should favor CSS in almost all cases (though there are probably edge cases). I'll take a look, but there could be opportunity to improve either way.

jpaims commented:
Design review:

| # | Issue | Category | Screenshot | 
| --- | ----------- | ----------- | --- | 
| 1 | Font-weight should be Semibold on selected (currently Bold) | Styling | ![Screenshot 2023-03-16 at 4 39 30 PM](https://user-images.githubusercontent.com/51726226/225775772-1f1e3cbf-4341-4898-88c9-7c211b9efef9.png) |
| 2 | Tab width should stay consistent between regular and selected weights. Currently there's some shifting going on.   | Sizing | 
| 3 | The hover underline should remain while the selected indicator is animating. Currently it disappears on press   | Styling | ![Screenshot 2023-03-16 at 4 40 14 PM](https://user-images.githubusercontent.com/51726226/225775844-cf3bc55d-4509-4824-ae8c-ef459c7c471f.png) |
| 4 | The text element should have 2px padding on the right and left for icon alignment purposes   | Sizing | ![Screenshot 2023-03-16 at 4 40 45 PM](https://user-images.githubusercontent.com/51726226/225775896-55b76b32-72be-48e2-8580-7e4834bf5631.png) |
| 5 | Vertical padding on vertical layout tabs is too large. Check [spec](https://www.figma.com/file/dK5AnDvvnSTWV9lduQWeDk/TabList?node-id=5107-14800&t=fxXC9r4gqPZlpav6-4) for spacing tokens  | Sizing | ![Screenshot 2023-03-16 at 4 41 26 PM](https://user-images.githubusercontent.com/51726226/225776015-8a4c69ee-effd-45d7-b41d-a7130801742f.png) |
| 6 | Keyboard arrow keys should move focus and Enter/Space selects the tab.  Currently arrow keys navigate and select the tab at the same time  | Accessibility |
| 7 | Check focus border style. Should be 2px stroke (--colorStrokeFocus2) combined with an inner 1px stroke (--colorStrokeFocus1 | Accessibility | ![Screenshot 2023-03-16 at 4 41 48 PM](https://user-images.githubusercontent.com/51726226/225776209-70ea6b29-c074-4bd5-8ea0-06be3f63ffe5.png) |
| 8 | Subtle appearance missing pressed state? | Interaction |
| 9 | Unselected tab text should use Foreground 2 instead of Foreground 1 | Styling |


**Questions:** 
Will we be adding icon support? 
Overflow functionality?

bearcat-msft commented:
> Design review:
> 
> #	Issue	Category	Screenshot
> 1	Font-weight should be Semibold on selected (currently Bold)	Styling	![Screenshot 2023-03-16 at 4 39 30 PM](https://user-images.githubusercontent.com/51726226/225775772-1f1e3cbf-4341-4898-88c9-7c211b9efef9.png)
> 2	Tab width should stay consistent between regular and selected weights. Currently there's some shifting going on.	Sizing	
> 3	The hover underline should remain while the selected indicator is animating. Currently it disappears on press	Styling	![Screenshot 2023-03-16 at 4 40 14 PM](https://user-images.githubusercontent.com/51726226/225775844-cf3bc55d-4509-4824-ae8c-ef459c7c471f.png)
> 4	The text element should have 2px padding on the right and left for icon alignment purposes	Sizing	![Screenshot 2023-03-16 at 4 40 45 PM](https://user-images.githubusercontent.com/51726226/225775896-55b76b32-72be-48e2-8580-7e4834bf5631.png)
> 5	Vertical padding on vertical layout tabs is too large. Check [spec](https://www.figma.com/file/dK5AnDvvnSTWV9lduQWeDk/TabList?node-id=5107-14800&t=fxXC9r4gqPZlpav6-4) for spacing tokens	Sizing	![Screenshot 2023-03-16 at 4 41 26 PM](https://user-images.githubusercontent.com/51726226/225776015-8a4c69ee-effd-45d7-b41d-a7130801742f.png)
> 6	Keyboard arrow keys should move focus and Enter/Space selects the tab. Currently arrow keys navigate and select the tab at the same time	Accessibility	
> 7	Check focus border style. Should be 2px stroke (--colorStrokeFocus2) combined with an inner 1px stroke (--colorStrokeFocus1	Accessibility	![Screenshot 2023-03-16 at 4 41 48 PM](https://user-images.githubusercontent.com/51726226/225776209-70ea6b29-c074-4bd5-8ea0-06be3f63ffe5.png)
> 8	Subtle appearance missing pressed state?	Interaction	
> 9	Unselected tab text should use Foreground 2 instead of Foreground 1	Styling	
> **Questions:** Will we be adding icon support? Overflow functionality?

Thanks James, I will address this feedback. Answer to questions - Icon support will be up to the dev implementing the control to add. It's low level of effort to add this. Overflow functionality – i have to review that in the design guidelines. I have not done anything to address overflow yet.

bearcat-msft commented:
Design review update: @jpaims I went through and addressed most of this feedback. Currently my implementation for the hover underline will only work in chromium browsers. Still looking for a fix for this.

| #   | Issue                                                                                                                                                                                     | Category      | Screenshot                                                                                                                                    | resolved? |
| --- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------- | --------------------------------------------------------------------------------------------------------------------------------------------- | --------- |
| 1   | Font-weight should be Semibold on selected (currently Bold)                                                                                                                               | Styling       | ![Screenshot 2023-03-16 at 4 39 30 PM](https://user-images.githubusercontent.com/51726226/225775772-1f1e3cbf-4341-4898-88c9-7c211b9efef9.png) | x         |
| 2   | Tab width should stay consistent between regular and selected weights. Currently there's some shifting going on.                                                                          | Sizing        |                                                                                                                                               | x         |
| 3   | The hover underline should remain while the selected indicator is animating. Currently it disappears on press                                                                             | Styling       | ![Screenshot 2023-03-16 at 4 40 14 PM](https://user-images.githubusercontent.com/51726226/225775844-cf3bc55d-4509-4824-ae8c-ef459c7c471f.png) | x         |
| 4   | The text element should have 2px padding on the right and left for icon alignment purposes                                                                                                | Sizing        | ![Screenshot 2023-03-16 at 4 40 45 PM](https://user-images.githubusercontent.com/51726226/225775896-55b76b32-72be-48e2-8580-7e4834bf5631.png) | x         |
| 5   | Vertical padding on vertical layout tabs is too large. Check [spec](https://www.figma.com/file/dK5AnDvvnSTWV9lduQWeDk/TabList?node-id=5107-14800&t=fxXC9r4gqPZlpav6-4) for spacing tokens | Sizing        | ![Screenshot 2023-03-16 at 4 41 26 PM](https://user-images.githubusercontent.com/51726226/225776015-8a4c69ee-effd-45d7-b41d-a7130801742f.png) | x         |
| 6   | Keyboard arrow keys should move focus and Enter/Space selects the tab. Currently arrow keys navigate and select the tab at the same time                                                  | Accessibility |                                                                                                                                               |           |
| 7   | Check focus border style. Should be 2px stroke (--colorStrokeFocus2) combined with an inner 1px stroke (--colorStrokeFocus1                                                               | Accessibility | ![Screenshot 2023-03-16 at 4 41 48 PM](https://user-images.githubusercontent.com/51726226/225776209-70ea6b29-c074-4bd5-8ea0-06be3f63ffe5.png) | x         |
| 8   | Subtle appearance missing pressed state?                                                                                                                                                  | Interaction   |                                                                                                                                               | x         |
| 9   | Unselected tab text should use Foreground 2 instead of Foreground 1                                                                                                                       | Styling       |                                                                                                                                               | x         |

eljefe223 commented on packages/web-components/src/tab-panel/tab-panel.styles.ts:
```diff
@@ -0,0 +1,13 @@
+    display: none;
+  }
+  :host {
+    display: block;
```
Use the `display` helper function

eljefe223 commented on packages/web-components/src/tab/tab.styles.ts:
```diff
@@ -0,0 +1,122 @@
+    color: ${colorNeutralForeground2};
+    fill: currentcolor;
+    grid-row: 1;
+    padding-top: ${spacingHorizontalM};
```
Lets shorthand this `padding: ${spacingHorizontalM} ${spacingHorizontalMNudge}`

eljefe223 commented on packages/web-components/src/tab/tab.styles.ts:
```diff
@@ -0,0 +1,122 @@
+    padding-right: ${spacingHorizontalMNudge};
+    border-radius: ${borderRadiusMedium};
+  }
+  :host::part(tab-content) {
```
Use a class instead.

eljefe223 commented on packages/web-components/src/tab/tab.styles.ts:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  /* adds hidden textContent to prevent shifting ui on bold / unbolding of text */
+  :host::part(tab-content)::after {
```
Class, class selectors are more performant than pseudo seelctors

eljefe223 commented on packages/web-components/src/tab/tab.styles.ts:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  :host([aria-selected='true'])::after {
+    background-color: ${colorCompoundBrandStroke};
```
Just curious why we are using psaudo elements here? Couldn't this be applied directly to the host?

bearcat-msft replied:
The background color here is for the the active indicator pseudo element. In order to not use a pseudo selector in this case we could add another dom element. Is that preferable to adding a pseudo element and selector?

eljefe223 replied:
Thats fine just trying to understand whats happening, Thanks!

eljefe223 commented on packages/web-components/src/tab/tab.styles.ts:
```diff
@@ -0,0 +1,122 @@
+    outline: none;
+  }
+
+  :host(:focus-visible) {
```
You have a duplicated selector here combine with the one on line 113

eljefe223 commented on packages/web-components/src/tab/tab.template.ts:
```diff
@@ -0,0 +1,14 @@
+  return html<T>`
+    <template slot="tab" role="tab" aria-disabled="${x => x.disabled}">
+      ${startSlotTemplate(options)}
+      <span part="tab-content"><slot></slot></span>
```
Add a class here

eljefe223 commented on packages/web-components/src/tabs/tabs.stories.ts:
```diff
@@ -0,0 +1,268 @@
+export const TabsSizeMedium = renderComponent(tabsSizeMedium).bind({});
+
+const tabsSizeLarge = html`
+  <div>
```
NIT do we need divs surrounding all these stories?

bearcat-msft replied:
removed

eljefe223 commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,258 @@
+
+export const styles = css`
+  :host {
+    display: grid;
```
display helper function

eljefe223 commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,258 @@
+
+  :host([disabled]) {
+    cursor: not-allowed;
+    color: ${colorNeutralForegroundDisabled} !important;
```
Are we running into specificity issues with this selector? Problem with important is its going to lock out any overrides.

bearcat-msft replied:
There is no specificity issue here. The reason important is added is so that this color can't be overridden, disabled always takes priority. (I though this was the way disabled was supposed to work)

eljefe223 replied:
Got it, its just not something we have we have done in other libraries but I guess if we feel this shouldn't ever be configurable from the consumer I'm fine with it. @chrisdholt for visibility

bearcat-msft replied:
Totally fine with whatever is the pattern for our WC library. In the past i have used !important to prevent users from overriding in certain cases, this seemed like appropriate use to me.

bearcat-msft replied:
I went ahead and removed the bang importants, since it seems like it isn't a pattern we're using.

eljefe223 commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,258 @@
+    color: ${colorNeutralForegroundDisabled} !important;
+  }
+
+  :host([disabled]) ::slotted(fluent-tab) {
```
The entire tabs component can be disabled?

bearcat-msft replied:
Yes, oddly, yes

eljefe223 commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,258 @@
+  }
+
+  // TODO: remove active-indicator styling when FAST Pr to remove active indicator is completed.
+  .active-indicator {
```
there is an Active indicator is an attribute you can just st to false, actually @chrisdholt Haven't we already removed this?

bearcat-msft replied:
You are right, it seems the .active-indicator style is not necessary. Removed!

eljefe223 commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,258 @@
+  }
+
+  :host([appearance='subtle']) ::slotted(fluent-tab:active) {
+    background-color: ${colorSubtleBackgroundPressed};
```
Just so I understand and I called it out before, we have a pseudo element we are targeting in some places and here we are not.

eljefe223 commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,258 @@
+    color: ${colorNeutralForeground1};
+  }
+
+  :host ::slotted(fluent-tab.animated)::after {
```
Ah is this why? To support the animation of the background?

bearcat-msft replied:
Yes, that's right. Here is a css-tricks article about how the animation here is working https://css-tricks.com/animating-layouts-with-the-flip-technique/ 

The css, pseudo elements and how the animation is applied with the javascript is very similar to Geoff Cox's implementation for the Fluent React TabList

eljefe223 commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,194 @@
+import { TabsAppearance, TabsSize } from './tabs.options.js';
+
+export const TabTokenNames = {
+  tabIndicatorOffset: '--tabIndicatorOffsetX',
```
DO we want these to be tokens or do we just want to assign css variables and use the `$fastController.addStyles`  @chrisdholt for his opinion.

bearcat-msft replied:
@chrisdholt do you have any comment here?

chrisdholt commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,194 @@
+  @attr
+  public size?: TabsSize = 'medium';
+
+  /**
+   * reserve-selected-tab-space
+   * Controls whether or not to make the text of the active tab bold. Bold text will cause a slight movement in the ui of the tabs. Defaults to false.
+   */
+  @attr({ attribute: 'reserve-selected-tab-space', mode: 'boolean' })
+  public reserveSelectedTabSpace?: boolean;
```
Should this really be an attribute and optional? This appears to just be "the way" via the latest in Fluent UI React v9: https://master--628d031b55e942004ac95df1.chromatic.com/?path=/docs/components-tablist--default

bearcat-msft replied:
yeah, okay, we can probably remove this

chrisdholt commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,194 @@
+   * Used to set the size of all the tab controls, which effects text size and margins. Three sizes: small, medium and large.
+   */
+  @attr
+  public size?: TabsSize = 'medium';
```
A couple notes here - if we're setting a value in code we should always use the actual value from the const object we've created. With that said, we should be setting these via CSS only and using tsdoc comments to enumerate what the default is. Setting this will reflect it back to the DOM which will clutter.

chrisdholt commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,194 @@
+   * setTabData
+   * sets the data from the active tab onto the class. used for making all the animation calculations for the active tab indicator.
+   */
+  private setTabData(): void {
```
this is a nit, but I'm not sure `set` makes sense here because you're simply calculating things..."set" makes it sound like you're doing work to actually set a value for the component to either provide in the DOM or to a user, neither of which are the case here.

bearcat-msft replied:
The purpose of this method is to save the active tab data onto the class at this.activeTabData. I thought anytime you have a statement where this.classField = someNewValue it is considered "setting."

chrisdholt commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    this.previousActiveTabData = this.activeTabData;
+    this.calculateAnimationProperties(tab);
```
What are we calculating if the tab itself is not selected?

bearcat-msft replied:
The tab is selected here. Here we have to calculate the ending location of the animation. But, probably these two statements should be moved below where activeTabOffset and scale are checked

chrisdholt replied:
I think I need to play with this a bit to understand the loop that's happening here.

bearcat-msft replied:
Geoff Cox referred me to this article - https://css-tricks.com/animating-layouts-with-the-flip-technique/ You may want to check it out.

chrisdholt commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,194 @@
+   * getSelectedTabPosition - gets the x or y coordinates of the tab
+   */
+  private getSelectedTabPosition(tab: Tab): number {
+    if (this.orientation === 'horizontal') {
```
Let's check this against the constant:

```ts
    if (this.orientation === TabsOrientation.horizontal) {
```

chrisdholt commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,194 @@
+   * getSelectedTabScale - gets the scale of the
+   */
+  private getSelectedTabScale(tab: Tab): number {
+    if (this.orientation === 'horizontal') {
```
Let's check this against the constant:

```ts
    if (this.orientation === TabsOrientation.horizontal) {
```

bearcat-msft commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,194 @@
+   */
+  private animationLoop(tab: Tab) {
+    // If selected, set the animation properties.
+    if (tab.ariaSelected && tab.ariaSelected === 'true') {
```
Where is the selected property? I don't see it on Tabs or Tab. Actually, I just realized I can remove this check anyway, since I'm only running the loop on the active tab. So the check is unecesarry.

We do need all this code in the loop. The animation loop only runs each time the active tab changes, and it does not run per tab.

chrisdholt commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,194 @@
+    document.documentElement.style.setProperty(TabTokenNames.tabIndicatorOffset, `${this.activeTabOffset}px`);
+  }
+
+  private setTabScaleCSSVar(newScale?: number) {
+    document.documentElement.style.setProperty(TabTokenNames.tabIndicatorScale, `${newScale || this.activeTabScale}`);
+  }
```
This is never called with a newScale value. Is that param necessary at all?

bearcat-msft replied:
oops!

bearcat-msft commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,194 @@
+      const parentRect = this.getBoundingClientRect();
+
+      this.activeTabData = {
+        id: activeTab.id,
```
to track the previously selected tab

chrisdholt replied:
Can we reasonably assume that if the data isn't the same we need to animate? All I see us checking is the existence of the id, not anything with regard to it's value. If we can remove this we can update our types to just be DOMRect.

bearcat-msft replied:
Yeah, that's a good point. That should work.

chrisdholt commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,185 @@
+  }
+
+  private setTabOffsetCSSVar() {
+    document.documentElement.style.setProperty(TabTokenNames.tabIndicatorOffset, `${this.activeTabOffset}px`);
```
What is `document.documentElement` in this case? Is it the root document? Is it the parent shadow DOM if nested? I'm a bit concerned on reaching outside the component to set this.

bearcat-msft replied:
I can update this to target the :host on this.styles. I was having trouble figuring out how to type the key for the css variable so I could use the enumerated object with the css variable names:

<img width="695" alt="image" src="https://user-images.githubusercontent.com/90428651/228334803-69730dbd-0a72-4eb0-b94b-4c75df0f616c.png">
<img width="920" alt="image" src="https://user-images.githubusercontent.com/90428651/228334903-5830e228-087f-47ef-a482-8ec1bec3f1b6.png">

bearcat-msft replied:
@eljefe223 do you know?

chrisdholt commented on packages/web-components/src/tab/tab.template.ts:
```diff
@@ -0,0 +1,14 @@
+  return html<T>`
+    <template slot="tab" role="tab" aria-disabled="${x => x.disabled}">
+      ${startSlotTemplate(options)}
+      <span class="tab-content"><slot></slot></span>
```
Are we only forking the template here to accommodate the "semibold" hack? :(

bearcat-msft replied:
yes... i needed a third place to add content programmatically. ::before - for the activeIndicator gray state (stays on while indicator animates). ::after - for the animating indicator. Then I needed a third place to add the semi-bolded content (this tab-content class). If you can think of a better way let me know. I wasn't over the moon about this solution, but it was the best one I could see.

chrisdholt commented on packages/web-components/src/tabs/tabs.stories.ts:
```diff
@@ -0,0 +1,235 @@
+
+const tabsDefault = html`
+  <fluent-tabs
+    orientation=${component => component.orientation}
+    appearance=${component => component.appearance}
+    ?disabled=${component => component.disabled}
+    size=${component => component.size}
+    activeid=${component => component.activeid}
```
I believe that we've been pretty consistently using `x => x.property`, we should probably leverage the same thing just for consistency.

bearcat-msft replied:
I saw that, and I am happy to change this to x => x.property for the sake of consistency. For the record though, in my opinion, it's not the best dev practice to give variables arbitrary names (unless it's a simple data type) and we should consider changing this project wide.

chrisdholt replied:
If you are particularly passionate about it and want to file an RFC, please do so. I'll note, for the stories, I think it's equally as clear as component. With that said, I don't really have a strong opinion besides the fact that we already use the format for templates and we should maintain that approach over verbosity IMO. For stories, don't really have a preference, but it would be odd to treat them differently.

chrisdholt commented on packages/web-components/src/tabs/tabs.stories.ts:
```diff
@@ -0,0 +1,235 @@
+    size=${component => component.size}
+    activeid=${component => component.activeid}
+  >
+    <fluent-tab id=${defaultTabIds[0]}> First Tab </fluent-tab>
+    <fluent-tab id=${defaultTabIds[1]}> Second Tab</fluent-tab>
+    <fluent-tab id=${defaultTabIds[2]}> Third Tab</fluent-tab>
+    <fluent-tab id=${defaultTabIds[3]}> Fourth Tab</fluent-tab>
+
+    <fluent-tab-panel id=${defaultTabIds[0] + '-panel'}>Tab One Content</fluent-tab-panel>
+    <fluent-tab-panel id=${defaultTabIds[1] + '-panel'}>Tab Two Content</fluent-tab-panel>
+    <fluent-tab-panel id=${defaultTabIds[2] + '-panel'}>Tab Three Content</fluent-tab-panel>
+    <fluent-tab-panel id=${defaultTabIds[3] + '-panel'}>Tab Four Content</fluent-tab-panel>
```
I'm going to nit on the stories here and the `id` usage. From a standards perspective, id's should be unique per document; these are all in the light DOM so it's kind of "bad practice" in that sense. From a tabs perspective, these _should_ get default id values if none is provided. Can we enumerate one set of id's for a story and then just not pass for the rest?

chrisdholt commented on packages/web-components/src/tabs/tabs.stories.ts:
```diff
@@ -0,0 +1,235 @@
+  argTypes: {
+    appearance: {
+      options: Object.values(TabsAppearanceValues),
+      defaultValue: 'transparent',
```
Let's be sure (for all the instances here) we use the actual const object to enumerate the default value.

eljefe223 commented on packages/web-components/src/tab-panel/tab-panel.styles.ts:
```diff
@@ -0,0 +1,13 @@
+import { spacingHorizontalM, spacingHorizontalMNudge } from '../theme/design-tokens.js';
+
+export const styles = css`
+  ${display('none')}
```
Duplicated displays :)

eljefe223 commented on packages/web-components/src/tab/tab.styles.ts:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  :host([aria-selected='true']),
+  :host([aria-selected='true'][disabled]) {
```
You probably dont need to specifically target disabled here unless its trying to set it own font-weight

eljefe223 commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,250 @@
+  :host([orientation='vertical']) ::slotted(fluent-tab)::before {
+    height: unset;
+    width: ${strokeWidthThicker};
+    margin-right: auto;
```
rtl? user `margn-inline-end`

chrisdholt commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,250 @@
+    color: ${colorNeutralForeground1};
+  }
+
+  :host ::slotted(fluent-tab.animated)::after {
```
Can we use an attribute rather than a class here?

chrisdholt commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,250 @@
+    padding-bottom: ${spacingVerticalSNudge};
+    padding-left: ${spacingHorizontalSNudge};
+    padding-right: ${spacingHorizontalSNudge};
+    border-radius: ${borderRadiusMedium};
```
If border-radius is consistent across all, can we reduce to just the primary selector all slotted tabs rather than duplicating between small/large (both are the same)

chrisdholt commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,250 @@
+    transform-origin: top;
+  }
+
+  :host([orientation='horizontal'][size='small']) ::slotted(fluent-tab)::after,
+  :host([orientation='horizontal'][size='small']) ::slotted(fluent-tab)::before,
+  :host([orientation='horizontal'][size='small']) ::slotted(fluent-tab:hover)::after {
+    right: ${spacingHorizontalSNudge};
+    left: ${spacingHorizontalSNudge};
+  }
+  :host([orientation='horizontal']) ::slotted(fluent-tab)::after,
+  :host([orientation='horizontal']) ::slotted(fluent-tab)::before,
+  :host([orientation='horizontal']) ::slotted(fluent-tab:hover)::after {
+    right: ${spacingHorizontalMNudge};
+    left: ${spacingHorizontalMNudge};
+  }
+  :host([orientation='horizontal'][size='large']) ::slotted(fluent-tab)::after,
+  :host([orientation='horizontal'][size='large']) ::slotted(fluent-tab)::before,
+  :host([orientation='horizontal'][size='large']) ::slotted(fluent-tab:hover)::after {
+    right: ${spacingHorizontalMNudge};
+    left: ${spacingHorizontalMNudge};
+  }
```
Horizontal is the default here, can we reduce the complexity of these selectors?

chrisdholt commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,186 @@
+import { Tab } from '../index.js';
+import { TabsAppearance, TabsSize } from './tabs.options.js';
+
+export const TabTokenNames = {
+  tabIndicatorOffset: '--tabIndicatorOffset',
+  tabIndicatorScale: '--tabIndicatorScale',
+} as const;
```
@bearcat-msft just my own $0.02, these aren't really used broadly. Having a constant for it seems like overkill. Not going to block, but I'm not sure we want these exposed to this extent.

eljefe223 replied:
@chrisdholt we could just assign variable using `$fastControler.addStyles`

bearcat-msft replied:
np, removed this.

eljefe223 commented on packages/web-components/src/tabs/tabs.stories.ts:
```diff
@@ -0,0 +1,235 @@
+    <fluent-tab> Third Tab</fluent-tab>
+    <fluent-tab> Fourth Tab</fluent-tab>
+
+    <fluent-tab-panel></fluent-tab-panel>
```
For testing should we add some content in the tab panels?

bearcat-msft replied:
The reason i did not add content in many cases in the tab panels is because the extra space and content seemed to clutter the page and detract from illustrating the different examples of the tabs layout (vertical, subtle, small, medium, large, ect).

eljefe223 commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,250 @@
+  :host([orientation='horizontal']) ::slotted(fluent-tab[aria-selected='false']:hover)::after {
+    height: ${strokeWidthThicker};
+    margin-top: auto;
+    transform-origin: left;
```
what about rtl here?

eljefe223 replied:
Just make sure it works in that writing mode

eljefe223 replied:
when document direction is set to `rtl`

bearcat-msft replied:
ah, okay, this won't be a problem here, since this only targets the hover indicator

bearcat-msft replied:
okay, checked. it's working in ltr and rtl

eljefe223 commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,250 @@
+    top: ${spacingVerticalSNudge};
+    bottom: ${spacingVerticalSNudge};
+  }
+  :host([orientation='vertical']) ::slotted(fluent-tab)::after,
```
combine `:host([orientation='vertical'][size='large']) ::slotted(fluent-tab)` here

eljefe223 commented on packages/web-components/src/tabs/tabs.styles.ts:
```diff
@@ -0,0 +1,250 @@
+    padding-bottom: ${spacingVerticalS};
+  }
+
+  :host([orientation='vertical'][size='small']) ::slotted(fluent-tab)::after,
```
combine `:host([orientation='vertical']) ::slotted(fluent-tab)` here

eljefe223 commented on packages/web-components/src/tabs/tabs.ts:
```diff
@@ -0,0 +1,186 @@
+   */
+  private animationLoop(tab: Tab) {
+    // remove the animated class so nothing animates yet
+    tab.classList.remove('animated');
```
Not to mention that if someone has a class on the tab the tab would also need to use `:classList` otherwise I'm not sure they would concatenate.

chrisdholt replied:
Responded offline but I don't see why this needs to be observable, so no need for the attr. This is a hook you're applying to manage styling/animation instead of a class.

eljefe223 replied:
Thats why data-attributes exist, its really just used for styling when these components are used together. Now if animate was an attribute that was meant to be controlled by the consumer ( something they could toggle on and off) depending on use case then an attribute would make sense. Even then the attribute would only be on the parent `tabs` component and then just assign `data-animate` to the individual `tab`

`HTML is designed with extensibility in mind for data that should be associated with a particular element but need not have any defined meaning. [data-* attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/data-*) allow us to store extra information on standard, semantic HTML elements without other hacks such as non-standard attributes, or extra properties on DOM.`

eljefe223 replied:
https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes

chrisdholt replied:
> The reason I questioned it was from a readability perspective. Should I make it clear at the top of the class that there is a data attribute that's part of the component's inner workings. Then I thought, maybe that's just overkill.

Yeah, I see these as internals just as I see the class usage as internals which didn't need to be documented.

eljefe223 commented on packages/web-components/src/tab/tab.styles.ts:
```diff
@@ -0,0 +1,111 @@
+
+  :host(:focus-visible) {
+    border-radius: ${borderRadiusSmall};
+    box-shadow: 0 0 0 2pt ${colorStrokeFocus2};
```
suppoed to be using `pt` here?

bearcat-msft replied:
fixed
